server/debugsession.go

```diff
@@ -0,0 +1,578 @@
+		ds.debugger = debugger
+		ds.backend.SetDebugger(debugger)
+
+		ds.send(&dap.InitializeResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+			Body: dap.Capabilities{
+				SupportsConfigurationDoneRequest: true,
+			},
+		})
+
+		ds.send(&dap.InitializedEvent{
+			Event: newDAPEvent("initialized"),
+		})
+
+	case *dap.SetBreakpointsRequest:
+		path := request.Arguments.Source.Path
+		location, err := pathLocation(path)
+		if err != nil {
+			ds.send(newDAPErrorResponse(
+				request.Seq,
+				request.Command,
+				dap.ErrorMessage{
+					Format: "cannot add breakpoints for path: {path}",
+					Variables: map[string]string{
+						"path": path,
+					},
+				},
+			))
+			break
+		}
+
+		ds.debugger.ClearBreakpointsForLocation(location)
+
+		requestBreakpoints := request.Arguments.Breakpoints
+
+		responseBreakpoints := make([]dap.Breakpoint, 0, len(requestBreakpoints))
+
+		for _, requestBreakpoint := range requestBreakpoints {
+			ds.debugger.AddBreakpoint(location, uint(requestBreakpoint.Line))
+
+			responseBreakpoints = append(
+				responseBreakpoints,
+				dap.Breakpoint{
+					Source: dap.Source{
+						Path: request.Arguments.Source.Path,
+					},
+					// TODO:
+					Verified: true,
+				},
+			)
+		}
+
+		ds.send(&dap.SetBreakpointsResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+			Body: dap.SetBreakpointsResponseBody{
+				Breakpoints: responseBreakpoints,
+			},
+		})
+
+	case *dap.LaunchRequest:
+		// TODO: only allow one program at a time
+
+		var args map[string]any
+		_ = json.Unmarshal(request.Arguments, &args)
+
+		programArg, ok := args["program"]
+		if !ok {
+			ds.send(newDAPErrorResponse(
+				request.Seq,
+				request.Command,
+				dap.ErrorMessage{
+					Format:   "Missing program",
+					ShowUser: true,
+				},
+			))
+			break
+		}
+		ds.code = programArg.(string)
+
+		stopOnEntryArg := args["stopOnEntry"]
+		ds.stopOnEntry, _ = stopOnEntryArg.(bool)
+
+		scriptID := emulator.ComputeScriptID([]byte(ds.code))
+		ds.scriptLocation = common.ScriptLocation(scriptID)
+
+		ds.launchRequested = true
+
+		ds.send(&dap.LaunchResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+		})
+
+		if ds.configurationDone && !ds.launched {
+			ds.run()
+		}
+
+	case *dap.ConfigurationDoneRequest:
+		ds.configurationDone = true
+
+		if ds.launchRequested && !ds.launched {
+			ds.run()
+		}
+
+		ds.send(&dap.ConfigurationDoneResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+		})
+
+	case *dap.ThreadsRequest:
+		ds.send(&dap.ThreadsResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+			Body: dap.ThreadsResponseBody{
+				Threads: []dap.Thread{
+					{
+						Id:   1,
+						Name: "Emulator",
+					},
+				},
+			},
+		})
+
+	case *dap.PauseRequest:
+		ds.debugger.RequestPause()
+
+		ds.send(&dap.PauseResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+		})
+
+	case *dap.NextRequest:
+		ds.step()
```

Comment:
I think this is step into, but comment on L222 makes it seem this is step-over.

Reply:
Yes, all stepping requests are just simple steps, there is no support for step into, step over, step out etc.